addressbook/interface/exchangeAutoCompleteResult/mivExchangeAutoCompleteResult.js

```diff
@@ -247,53 +244,52 @@ mivExchangeAutoCompleteResult.prototype = {
      * persistent storage as well.
      */
     //void removeValueAt(in long rowIndex, in boolean removeFromDb);
-    removeValueAt: function _removeValueAt(aRowIndex, removeFromDb) {},
+    removeValueAt: function _removeValueAt(aRowIndex, removeFromDb) {
+        var localid = this._idcards.splice(aRowIndex, 1);
+        delete this._cards[localid];
+    },
 
     //void addResult(in mivExchangeAbCard aCard);
     addResult: function _addResult(aCard) {
         // First check if this card is not already in the list
         var cardExists = false;
-        if (this._cards) {
-            for (var card of Object.values(this._cards)) {
-                if (card.localId == aCard.localId) {
 
-                    cardExists = true;
-                    break;
-                }
-            }
+        if (this._cards
+            && this._cards[aCard.localId]) {
+                cardExists = true;
         }
+
         if (!cardExists) {
-            //dump("addResult:"+aCard.displayName+", primaryEmail:"+aCard.primaryEmail+", length:"+this._cards.length+"\n");
-            //			if ((aCard.primaryEmail != "") && (aCard.primaryEmail.indexOf("@") > -1)) {
-            if (((aCard.primaryEmail != "") && (aCard.primaryEmail.indexOf("@") > -1)) || (aCard.isMailList)) {
-                this._cards.push(aCard);
+            if ((aCard.primaryEmail != "" && aCard.primaryEmail.indexOf("@") > -1)
```

Comment:
The first condition is redundant since `indexOf("@")` will never be greater `-1` for an empty string. Probably should be rearranged to first test for `isMailList` like in code higher up.

Reply:
Ok, I've read again the code, and maybe this suggestion is better ?

```javascript
        // Before really adding the result, check the card is a mailing list
        // and, otherwise, check it's primary email has at least a "@"
        if (aCard.isMailList
            || (aCard.primaryEmail && aCard.primaryEmail.indexOf("@") > -1)) {
            this._cards[aCard.localId] =  aCard;
            this._idcards.push(aCard.localId);
        }
```

Reply:
Assuming `aCard.primaryEmail` can be null or undefined, this is ok.

The original condition however was against `""` (empty string), so if in this case `aCard.primaryEmail` can be assumed to never be null or undefined, you can also get rid of the `aCard.primaryEmail` condition since `aCard.primaryEmail.indexOf("@") > -1` will do everything that is needed (i.e. `indexOf(...)` will always return `-1` for an empty string).